kobotoolbox/kpi
# feat(projectHistoryLogs): create project history logs when related objects change TASK-944

### 🗒️ Checklist

1. [x] run linter locally
2. [x] update all related docs (API, README, inline, etc.), if any
3. [x] draft PR with a title `<type>(<scope>)<!>: <title> TASK-1234`
4. [x] tag PR: at least `frontend` or `backend` unless it's global
5. [x] fill in the template below and delete template comments
6. [x] review thyself: read the diff and repro the preview as written
7. [x] open PR & confirm that CI passes
8. [x] request reviewers, if needed
9. [ ] delete this section before merging

### 📣 Summary
Create project history logs for changes to associated media files, rest services, and imported data.

### 📖 Description
Log changes whenever someone:
1. adds/deletes/modifies a REST service
2. adds/deletes a media attachment (project, not submission)
3. dis/connects to another project
4. modifies fields imported from another project

### 👀 Preview steps
1. ℹ️ Log in as superuser and create a survey
2. Deploy the survey
3. Go to Settings > Media and upload a small jpg:
![tiara](https://github.com/user-attachments/assets/c63cf88c-aa94-4bf7-ae6d-8d1da30c8f28)
5. Go to `localhost/api/v2/audit-logs/?q=log_type:project-history AND metadata__asset_uid:<asset_uid>`
6. 🟢The latest log should look like 
```
{
    "app_label":"kpi",
    "model_name":"asset",
    "user":"http://localhost/api/v2/users/<username>/?format=json",
    "user_uid":"<your user uid>",
    "username":"<your username>",
    "action":"add-media",
    "metadata":{
        "source":"<browser><OS>",
        "asset_uid":"<asset-uid>",
        "asset-file":{
            "uid":"<uid>",
            "filename":"tiara.jpg",
            "md5_hash":"md5:e9ad5b5b0b66e3c063c4d067bfe54ca4",
            "download_url":"<some_url>"
        },
        "ip_address":"<your ip>",
        "log_subtype":"project"
    },
    "date_created":"20
```
7. Verify the download url is correct by going to localhost/<download_url>
8. Remove the picture from the project
9. Reload the endpoint. You should see a new log with the same metadata and `action: delete-media`
10. Go to Settings > Rest Services > Register a new service
11. Add a new service named 'Test' with the endpoint 'http://www.google.com'. Keep it enabled.
12. 🟢 Reload the endpoint. You should see a new log with `action: register-service` and the following in `metadata.hook`:
```
{
      "uid":"<uid>",
      "active":true,
      "endpoint":"https://www.google.com"
},
```
12. Pause the hook by editing it and de-selecting 'Enabled'
14. 🟢 Reload the endpoint. You should see a new log with `action: register-service` and the following in `metadata.hook`:
 ```
{
      "uid":"<uid>",
      "active":false,
      "endpoint":"https://www.google.com"
},
```
15. Create and deploy a new survey named 'Test connect' with questions `q1` and `q2`
16. Go to Settings > Connect Project and enable data sharing (all questions)
17. Go back to the original project and go to Settings > Connect Project
18. Select a different project to import data from > 'Test connect' > Import > Import all fields
19. 🟢Reload the endpoint. You should see a new log with `action: connect-project` and the following in `metadata.paired_data`:
 ```
{
      "source_name":"Test connect",
      "source_uid":"<test connect uid>",
      "fields":["q1", "q2"]
},
```
20. Go back to Settings > Connect Project and click the Gear icon next to 'Test connect' (under Imported)
21. Deselect `q1` from the imported questions and hit `Accept`
22. 🟢Reload the endpoint. You should see a new log with `action: modify-imported-fields` and the following in `metadata.paired_data`:
 ```
{
      "source_name":"Test connect",
      "source_uid":"<test connect uid>",
      "fields":["q2"]
},
```
23. Disconnect the projects by clicking the trash can icon next to 'Test connect' 
24. 🟢 Reload the endpoint. You should see a new log with `action: disconnect-project` and the following in `metadata.paired_data`:
 ```
{
      "source_name":"Test connect",
      "source_uid":"<test connect uid>",
      "fields":["q2"]
},
```

### 💭 Notes
Updates the AuditLoggedViewSet code to allow fields to be put in the `initial_data` and/or `updated_data` dictionaries under a different label if desired. If logged_fields has a tuple (k,v) instead of a field, the relevant dictionary will store `{'k': object.v}` rather than `{v: object.v}`. 
Updates the HookViewSet, AssetFileViewSet, and PairedDataViewsets to inherit from the AuditLoggedViewSet base class, adding the necessary fields from each to the `logged_fields` as well as the object id of the related asset.
Add a method in ProjectHistoryLog for dealing with requests that come from these kinds of related-object view sets, where it just takes everything in the `updated_data` or `initial_data` dictionary (except the `object_id`, which it uses as its own `object_id`) and puts it in the metadata under a given label. 
Also refactors the overall `create_from_request` delegation method to use a dictionary of `<view_name> : method` rather than a whole bunch of if-elses. 
Unrelated to other things, changes calls to the deprecated method `assertEquals` to `assertEqual` to get rid of warnings.

BTW @noliveleger in order to make PRs less confusing I'll keep going the way I've been doing it then do an alphabetizing PR at the very end.
